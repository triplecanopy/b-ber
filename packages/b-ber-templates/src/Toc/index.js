import File from 'vinyl'
import { Html } from '@canopycanopycanopy/b-ber-lib'
import { getTitle } from '@canopycanopycanopy/b-ber-lib/utils'

class Toc {
  static body() {
    return new File({
      contents: Buffer.from(
        `<nav id="toc" epub:type="toc"><h2>Table of Contents</h2>{% body %}</nav>`
      ),
    })
  }

  static item(data) {
    const url = `${data.relativePath}.xhtml`
    return `<a href="${url}">${Html.escape(getTitle(data))}</a>`
  }

  static items(data) {
    return `
      <ol>
        ${data.reduce((acc, curr) => {
          if (curr.in_toc === false) return acc

          return acc.concat(`
            <li>
              ${Toc.item(curr)}
              ${curr.nodes && curr.nodes.length ? Toc.items(curr.nodes) : ''}
            </li>
          `)
        }, '')}
      </ol>
  `
  }
}

export default Toc
