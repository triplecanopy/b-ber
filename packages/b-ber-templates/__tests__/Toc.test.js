/* eslint-disable camelcase */
import fs from 'fs-extra'
import Toc from '../src/Toc'

jest.mock('@canopycanopycanopy/b-ber-lib/State', () => ({
  spine: {
    frontMatter: new Map(),
  },
}))

afterAll(() => fs.remove('_project'))

describe('templates.Toc', () => {
  it('creates an anchor element', () => {
    const node = { in_toc: true, relativePath: 'test', title: 'Test' }
    expect(Toc.item(node)).toMatchSnapshot()
  })

  it('creates an ordered list', () => {
    const node = {
      in_toc: true,
      relativePath: 'l1',
      title: 'Test',
      nodes: [
        {
          in_toc: true,
          relativePath: 'l2',
          title: 'Test',
          nodes: [
            { in_toc: true, relativePath: 'l3', title: 'Test', nodes: [] },
          ],
        },
      ],
    }
    expect(Toc.item(node)).toMatchSnapshot()
  })

  it('creates a nav element', () => {
    expect(Toc.body().contents).toMatchSnapshot()
  })

  it('creates a nested list', () => {
    const node = { in_toc: true, relativePath: 'test', title: 'Test' }
    const nodes = [
      { in_toc: true, relativePath: 'test-child', title: 'Test Child' },
    ]
    const data = [
      {
        in_toc: false,
      },
      {
        ...node,
        nodes,
      },
    ]
    expect(Toc.items(data)).toMatchSnapshot()
  })
})
