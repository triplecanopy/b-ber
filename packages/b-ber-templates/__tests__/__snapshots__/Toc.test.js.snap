// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`templates.Toc creates a nav element 1`] = `
Object {
  "data": Array [
    60,
    110,
    97,
    118,
    32,
    105,
    100,
    61,
    34,
    116,
    111,
    99,
    34,
    32,
    101,
    112,
    117,
    98,
    58,
    116,
    121,
    112,
    101,
    61,
    34,
    116,
    111,
    99,
    34,
    62,
    60,
    104,
    50,
    62,
    84,
    97,
    98,
    108,
    101,
    32,
    111,
    102,
    32,
    67,
    111,
    110,
    116,
    101,
    110,
    116,
    115,
    60,
    47,
    104,
    50,
    62,
    123,
    37,
    32,
    98,
    111,
    100,
    121,
    32,
    37,
    125,
    60,
    47,
    110,
    97,
    118,
    62,
  ],
  "type": "Buffer",
}
`;

exports[`templates.Toc creates a nested list 1`] = `
"
      <ol>
        
            <li>
              <a href=\\"text/test.xhtml\\">Test</a>
              
      <ol>
        
            <li>
              <a href=\\"text/test-child.xhtml\\">Test Child</a>
              
            </li>
          
      </ol>
  
            </li>
          
      </ol>
  "
`;

exports[`templates.Toc creates an anchor element 1`] = `"<a href=\\"text/test.xhtml\\">Test</a>"`;

exports[`templates.Toc creates an ordered list 1`] = `"<a href=\\"text/l1.xhtml\\">Test</a>"`;
